### DISCUSSION TITLE
My silent expression keymap is not treated as such

### DISCUSSION THREAD

**Discussion by martin-braun** [NvChad/NvChad#1623] [Q&A] [answered]
I'm currently trying to keymap coc.nvim. To start simple I want to apply a mapping from the [README of coc.nvim](https://github.com/neoclide/coc.nvim/blob/master/README.md).

I try to translate these lines to NvChad: 

```lua
local keyset = vim.keymap.set
-- ...
local opts = {silent = true, noremap = true, expr = true, replace_keycodes = false}
-- ...
keyset("i", "<cr>", [[coc#pum#visible() ? coc#pum#confirm() : "\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>"]], opts)
```

Here is my attempt: 

```lua
-- ...
i = {
    -- ...
    ["<Cr>"] = { [[coc#pum#visible() ? coc#pum#confirm() : "\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>"]], "confirm selection", { silent = true, noremap = true, expr = true, replace_keycodes = false } },
-- ...
}
```

This line should check if the suggestion panel is up, if so, it should select the selection, otherwise it should just pass the carriage return forward.

I know that this plugin is not supported officially, but I think the issue is either simple or I spotted a bug, because it will fail the expression and print

```
coc#pum#visible() ? coc#pum#confirm() : "\\
\
```

into my file. Even when I replace `\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>` with `<CR>` it will just print `coc#pum#visible()`... into my file, so I guess NvChad is not treating my mapping as expression?

Could you please lead me into the right direction @siduck? I got the feeling I'm doing an obvious mistake here.

Thanks.

**Comment by siduck** [comment, marked as answer]
@martin-braun does it work with vim.keymap.set?

Also join any of our communities 

![image](https://user-images.githubusercontent.com/59060246/194678270-5b6d42d5-09c0-4d1d-abda-21683bfb6a0d.png)

Many can help you in my absence :D

**Reply by martin-braun** [reply]
Good idea, will do that, still need to report that things work when I decided to just paste the example Lua config from coc.nvim into my chadrc.lua: 

<img width="651" alt="Screen Shot 2022-10-08 at 1 48 08 PM" src="https://user-images.githubusercontent.com/24259245/194705963-75146712-b439-4e07-a27d-33a1cb9c8e99.png">

I loaded the release branch from coc.nvim with `["neoclide/coc.nvim"] = { branch = 'release' },` and I pasted all the Lua content from the README at the bottom of chadrc.lua.

I'd prefer to get this right and integrate it into NvChad. The config conversion does not work for me, though.

**Reply by martin-braun** [reply]
@siduck Been slowly moving my keymaps to NvChad. All my keymaps work, except for the keymaps that involve `expr = true`, those get just printed in the editor instead of executed. 

For instance, see this simple keymap in insert mode: 

```lua
["<C-Cr>"] = { 'coc#refresh()', "show suggestions", { silent = true, expr = true } },
```

Even without coc.nvim installed, you should be able to replicate and confirm this behavior when pressing CTRL+CR: 

![2022-10-08 3 00 01 PM](https://user-images.githubusercontent.com/24259245/194708804-4f90f6a8-1e15-48da-9ff3-35e5ea213c8e.gif)

My understanding is that Neovim should execute this expression and show suggestions of coc.nvim or at least an error message in the status line when coc.nvim is missing.

EDIT: To verify, here is the normal Lua bit that works as expected: 

```lua
vim.keymap.set("i", "<C-Cr>", "coc#refresh()", {silent = true, expr = true})
```

**Reply by martin-braun** [reply]
Okay, messed up here, this works:

```lua
["<C-Cr>"] = { 'coc#refresh()', "show suggestions", opts = { silent = true, expr = true } },
```

Should've read the documentation better. The community on Telegram / Matrix is awesome. Thanks! =)
